### PR TITLE
Autorouter fixes for 1860, 1862, 1873, 1849, plus probably more

### LINKED ISSUE #6100: Auto-routing crashes on 1849
https://18xx.games/game/47530?action=613

h
message: "Array can't be coerced into Numeric"
name: "TypeError"
stack: "TypeError: Array can't be coerced into Numeric\n    at Object.t.send2 (https://18xx.games/assets/deps.js:1:16269)\n    at t.send (https://18xx.games/assets/deps.js:1:16107)\n    at Function.u (https://18xx.games/assets/deps.js:5:14231)\n    at Number.He (https://18xx.games/assets/deps.js:5:9559)\n    at https://18xx.games/assets/deps.js:9:60961\n    at Number.u (https://18xx.games/assets/deps.js:9:61216)\n    at Number.g (https://18xx.games/assets/deps.js:9:120536)\n    at H (https://18xx.games/assets/main.js:2:217435)\n    at h.E (https://18xx.games/assets/main.js:2:219912)\n    at ie (https://18xx.games/assets/main.js:1:95758)"

### PR DESCRIPTION
**Pull request by roseundy**
Modified autorouter to handle intra-tile paths correctly (1860, 18Mex, 18Texas, 18MS).
Modified 1860, 1862 and 1873 to deal with the autorouter path pruning algorithm (i.e. skip them if route.routes is empty).
Modified Part::RevenuCenter to bail gracefully (needed for 1862 offboards with the autorouter).
Modifed Route to display more useful error information when checking chains.

Fixes #6100

### CODE REVIEW

File: lib/engine/auto_router.rb [hunk 1]
```diff
@@ -63,12 +67,24 @@ def compute(corporation, **opts)
...
+            else
+              left = a
+              right = b
+            end
+            complete.call
+          end
+
           paths.each do |path|
             chain << path
             a, b = path.nodes
```
Comment (tobymao) [inline lib/engine/auto_router.rb:83]: maybe this can be cleaned up with 

assign.call(a, b) if a || b

File: lib/engine/game/g_1860/game.rb [hunk 2]
```diff
@@ -1683,6 +1683,7 @@ def visit_route(ridx, intersects, visited)
         # all routes must intersect each other
         def check_intersection(routes)
           actual_routes = routes.reject { |r| r.chains.empty? }
+          return unless actual_routes
```
Comment (tobymao) [inline lib/engine/game/g_1860/game.rb]: actual_routes will never be false, arrays are always truthy

File: lib/engine/part/revenue_center.rb [hunk 3]
```diff
@@ -57,6 +57,7 @@ def route_base_revenue(phase, train)
         return (@revenue[:diesel]) if train.name.upcase == 'D' && @revenue[:diesel]
 
         phase.tiles.reverse_each { |color| return (@revenue[color]) if @revenue[color] }
+        1
```
Comment (tobymao) [inline lib/engine/part/revenue_center.rb:60]: maybe default to 0?

File: lib/engine/auto_router.rb (hunk 1 above)
Comment (roseundy) [reply lib/engine/auto_router.rb:83]: No, that will still call if a and b are true.

File: lib/engine/game/g_1860/game.rb (hunk 2 above)
Comment (roseundy) [reply lib/engine/game/g_1860/game.rb]: Jinx. You owe me a soda.

File: lib/engine/part/revenue_center.rb (hunk 3 above)
Comment (roseundy) [reply lib/engine/part/revenue_center.rb:60]: Sure.

File: lib/engine/auto_router.rb (hunk 1 above)
Comment (tobymao) [reply lib/engine/auto_router.rb:83]: you'll have to rewrite the assigns methods

Comment (roseundy) [reply lib/engine/auto_router.rb:83]: Yep.

File: lib/engine/auto_router.rb [hunk 4]
```diff
@@ -47,15 +47,33 @@ def compute(corporation, **opts)
...
             chains << { nodes: [left, right], paths: chain }
+            last_left = left
+            last_right = right
             left, right = nil
             chain = []
           end
 
-          assign = lambda do |n|
-            if !left
+          assign_both = lambda do |a, b|
```
Comment (tobymao) [inline lib/engine/auto_router.rb:61]: can we delete this lambda?

Comment (roseundy) [reply lib/engine/auto_router.rb:61]: OK. It will make the other larger, but sure.

**Review by tobymao — APPROVED** [review]

**Pull request merged**